d2/src/test/java/com/linkedin/d2/balancer/strategies/framework/LoadBalancerStrategyTestRunner.java

```diff
@@ -0,0 +1,235 @@
+      Map<URI, Integer> lastRequestCountMap, Map<URI, Integer> currentRequestCountMap, Map<URI, Integer> callCountMap, Map<URI, Long> latencySumMap)
+  {
+    _strategy = strategy;
+    _serviceName = serviceName;
+    _numIntervals = numIntervals;
+    _requestsManager = requestsManager;
+    _trackerClients = trackerClients;
+    _clockedExecutor = clockedExecutor;
+
+    _currentErrorMap = currentErrorMap;
+    _lastRequestCountMap = lastRequestCountMap;
+    _currentRequestCountMap = currentRequestCountMap;
+    _callCountMap = callCountMap;
+    _latencySumMap = latencySumMap;
+  }
+
+  public List<TrackerClient> getTrackerClients()
+  {
+    return _trackerClients;
+  }
+
+  public URI getUri(int index)
+  {
+    return _trackerClients.get(index).getUri();
+  }
+
+  /**
+   * Get points of the partitions at the end of the test
+   * @return The URI to points map
+   */
+  public Map<URI, Integer> getPoints()
+  {
+    // TODO: Add other strategy types here
+    if (_strategy instanceof DegraderLoadBalancerStrategyV3)
+    {
+      return ((DegraderLoadBalancerStrategyV3) _strategy).getPointsMap(DEFAULT_PARTITION_ID);
+    }
+    // We should not get points if the strategy is not using hash ring
+    return new HashMap<>();
+  }
+
+  /**
+   * Get the points history for past intervals
+   */
+  public Map<URI, List<Integer>> getPointHistory()
+  {
+    return _pointHistoryMap;
+  }
+
+  /**
+   * Get the average latency for all the hosts during the test
+   * @return the average latency for all the hosts during the test
+   */
+  public double getAvgLatency()
+  {
+    long latencySum = 0;
+    int callCountTotal = 0;
+
+    for (URI uri : _callCountMap.keySet())
+    {
+      callCountTotal += _callCountMap.getOrDefault(uri, 0);
+      latencySum += _latencySumMap.getOrDefault(uri, 0L);
+    }
+
+    return latencySum / callCountTotal;
+  }
+
+  /**
+   * Run the test until it finishes
+   */
+  public void runWait()
+  {
+    Future<Void> running = run();
+    if (running != null)
+    {
+      try
+      {
+        running.get();
+      }
+      catch (InterruptedException | ExecutionException e)
+      {
+        _log.error("Test running interrupted", e);
+      }
+    }
+  }
+
+  /**
+   * Run the mocked test for the given intervals, each interval is scheduled to be run at the fixed interval time
+   */
+  private Future<Void> run()
+  {
+    _clockedExecutor.scheduleWithFixedDelay(new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        runInterval();
+      }
+    }, 10, LoadBalancerStrategyTestRunnerBuilder.INTERVAL_IN_MILLIS, TimeUnit.MILLISECONDS);
+    return _clockedExecutor.runFor(LoadBalancerStrategyTestRunnerBuilder.INTERVAL_IN_MILLIS * _numIntervals);
+  }
+
+  /**
+   * Execute one interval with the given request count
+   */
+  private void runInterval()
+  {
+    int currentIntervalIndex = (int) (_clockedExecutor.currentTimeMillis() / LoadBalancerStrategyTestRunnerBuilder.INTERVAL_IN_MILLIS);
```

Comment:
Do we know the clockedExecutor timer starts with 0? If not it should be currentTimeMillis - previousTimeStamp.

Reply:
Yes, each test will always create a new instance of the testRunner with a new clockedExecutor, we won't reuse the same instance again in a different test. Each test will run the test start from time 0.